darwin/CpuFreq.c

```diff
@@ -0,0 +1,176 @@
+     * Frequencies for the "E" cluster type are stored in voltage-states1,
+     * frequencies for the "P" cluster type are stored in voltage-states5.
+     * This seems to be hardcoded.
+     */
+    const CFStringRef voltage_states_key_per_cluster[CPUFREQ_NUM_CLUSTER_TYPES] = {CFSTR("voltage-states1"), CFSTR("voltage-states5")};
+
+    io_registry_entry_t pmgr_registry_entry = IORegistryEntryFromPath(kIOMainPortDefault, "IODeviceTree:/arm-io/pmgr");
+    if (pmgr_registry_entry == MACH_PORT_NULL) {
+        return -1;
+    }
+    for (size_t i = 0U; i < CPUFREQ_NUM_CLUSTER_TYPES; i++) {
+        CFDataRef voltage_states_ref =
+            (CFDataRef) IORegistryEntryCreateCFProperty(pmgr_registry_entry, voltage_states_key_per_cluster[i], kCFAllocatorDefault, 0U);
+        if (voltage_states_ref == NULL) {
+            IOObjectRelease(pmgr_registry_entry);
+            return -1;
+        }
+
+        CpuFreqPowerStateFrequencies* cluster_frequencies = &data->cpu_frequencies_per_cluster_type[i];
+        cluster_frequencies->num_frequencies = CFDataGetLength(voltage_states_ref) / 8;
+        cluster_frequencies->frequencies = xCalloc(cluster_frequencies->num_frequencies, sizeof(double));
+        const uint8_t* voltage_states_data = CFDataGetBytePtr(voltage_states_ref);
+        for (size_t j = 0U; j < cluster_frequencies->num_frequencies; j++) {
+            uint32_t freq_value;
+            memcpy(&freq_value, voltage_states_data + j * 8, 4);
+            cluster_frequencies->frequencies[j] = (65536000.0 / freq_value) * 1000000;
+        }
+        CFRelease(voltage_states_ref);
+    }
+    IOObjectRelease(pmgr_registry_entry);
+
+
+    /* Create subscription for CPU performance states */
+    CFMutableDictionaryRef channels = IOReportCopyChannelsInGroup(CFSTR("CPU Stats"), CFSTR("CPU Core Performance States"), NULL, NULL);
+    if (channels == NULL) {
+        return -1;
+    }
+
+    data->subscribed_channels = NULL;
+    data->subscription = IOReportCreateSubscription(NULL, channels, &data->subscribed_channels, 0U, NULL);
+
+    CFRelease(channels);
+
+    if (data->subscription == NULL) {
+        return -1;
+    }
+
+    data->prev_samples = NULL;
+
+    return 0;
+}
+
+void CpuFreq_update(CpuFreqData* data) {
+    CFDictionaryRef samples = IOReportCreateSamples(data->subscription, data->subscribed_channels, NULL);
+    if (samples == NULL) {
+        return;
+    }
+
+    if (data->prev_samples == NULL) {
+        data->prev_samples = samples;
+        return;
+    }
+
+    /* Residency is cumulative, we have to diff two samples to get a current view */
+    CFDictionaryRef samples_delta = IOReportCreateSamplesDelta(data->prev_samples, samples, NULL);
+
+    /* Iterate through performance state residencies. Index 0 is the idle residency, index 1-n is the per-performance state residency. */
+    __block uint32_t cpu_i = 0U;
+    IOReportIterate(samples_delta, ^(IOReportSampleRef ch) {
```

Comment:
As that's not strictly C99 conforming code, I assume that's intermixed ObjC stuff thrown in for good measure?

Reply:
Not ObjC, rather an Apple extension for [Blocks](https://en.wikipedia.org/wiki/Blocks_(C_language_extension)#:~:text=Blocks%20are%20a%20non%2Dstandard,create%20closures%20within%20these%20languages.). We must use it, since that's what the API expects for aa a callback.